### ISSUE TITLE
Inconsistent DNS resolution

### ISSUE THREAD

**Issue by Jocaru** [wader/static-ffmpeg#304]
Hello,

Sometimes when I try to use `ffmpeg` to stream video from a URL I get DNS errors, `ffmpeg` not being able to resolve the hostname, I did a `-v trace` but it shows just the same information without the option enabled.

```
[https @ 0x7fc0c12fd300] header='HTTP/1.1 302 Found'
[https @ 0x7fc0c12fd300] http_code=302
[https @ 0x7fc0c12fd300] header='Server: ####/0.1'
[https @ 0x7fc0c12fd300] header='Date: Wed, 29 Mar 2023 10:21:24 GMT'
[https @ 0x7fc0c12fd300] header='Connection: Close'
[https @ 0x7fc0c12fd300] header='Content-Length: 0'
[https @ 0x7fc0c12fd300] header='Location: #####'
[https @ 0x7fc0c12fd300] header='Access-Control-Allow-Headers: *'
[https @ 0x7fc0c12fd300] header='Access-Control-Expose-Headers: X-TCDN'
[https @ 0x7fc0c12fd300] header='X-TCDN: Host: ####.1, Type: m'
[https @ 0x7fc0c12fd300] header=''
[tcp @ 0x7fc0c12fd500] Failed to resolve hostname REDACTED.com: Try again
[dash @ 0x7fc0c15b2840] Failed to open an initialization section
[dash @ 0x7fc0c15b2840] Error when loading first fragment of playlist
[AVIOContext @ 0x7fc0c1501340] Statistics: 32768 bytes read, 0 seeks
[AVIOContext @ 0x7fc0c143da80] Statistics: 32768 bytes read, 0 seeks
[AVIOContext @ 0x7fc0c1437380] Statistics: 32768 bytes read, 0 seeks
[AVIOContext @ 0x7fc0c12ed280] Statistics: 32768 bytes read, 0 seeks
[AVIOContext @ 0x7fc0c1556b40] Statistics: 31938 bytes read, 0 seeks
https://REDACTED: I/O error
```

This happened after it being able to resolve several domains before reaching this point in the same `ffmpeg` execution.

If I immediately try again, it may or may not work, it usually does at the second attempt.

Is there any explanation for this? Using `dig` inside the same container I never have any problem resolving the same domains. Does `ffmpeg` rely on OS for DNS resolution or does it use its own system?

**Comment by wader** [comment]
Hi, yes ffmpeg relay on the OS, musl libc in case of static-ffmpeg, to resolve. Do you know dig has it's own DNS implementation or uses libc?

The only thing that comes to mind is https://github.com/wader/static-ffmpeg#i-see-name-does-not-resolve-errors-for-hosts-that-should-resolve that musl libc does not support DNS over TCP. When you do dig do you see a lot of answers sometimes? or maybe it's possible to see if dig swaps to using TCP?

**Comment by binoculars** [comment]
Just came across this recently: https://martinheinz.dev/blog/92

**Comment by wader** [comment]
Quick look at the bind-utils code it do seem to have it's own DNS implementation, which makes sense i guess if it's part of bind :)

**Comment by wader** [comment]
@Jocaru did you figure something more about? seems to be DNS?

**Comment by Jocaru** [comment]
@wader no, I still have the problem sometimes. I have read the link that @binoculars sent and that may be the problem I guess. The domains I use in my `ffmpeg` command are CNAMEs to another one and there are also redirections and that may be the problem (because of `musl`). I guess I should just use Debian or Ubuntu instead of Alpine.

I'm also just using your `ffmpeg` executable outside of your container (in another container). I did a quick test in a Ubuntu container but couldn't execute it. I ended up using yours because it has all libraries that I need (I must use the `-cenc_decryption_key` option).

I don't know if it's much asking, but could you do an Ubuntu/Debian image?

**Comment by wader** [comment]
Aha i see. Problem with building glibc based static executables is that they will not be truely static and might up loading dynamic things, if i remember correctly the NSS stuff i dynamic. So would probably have to build per distro etc. 

I do know of stripped down "forks" of this Dockerfile that build and run on ubuntu, is a bit of work to do but possible.

Another alternative might be to build with the musl DNS patches or build with master version of musl somehow.

A second alternative could be to run thru non-musl based http proxy or let the program that runs ffmpeg do DNS resolve and rewrite the URL to use IP:s (might require some host header trickery?), both ways hacky but might work.

How urgen is it to be fixed? i guess it will be some months until the DNS fixes are in alpine stable.

**Comment by Jocaru** [comment]
Not urgent at all, I'm just using that for a personal project :)

**Comment by wader** [comment]
Okok! i thought `-cenc_decryption_key` was quite old but was added in 5.1 https://github.com/FFmpeg/FFmpeg/commit/8b64d8d9aa7fede94ab531b7be3f73cd2f3c448c so maybe you can use some other image like https://github.com/linuxserver/docker-ffmpeg that seem be ubuntu-based and have 5.1+.

**Comment by wader** [comment]
... and of course come back to wader/static-ffmpeg once alpine uses the new musl version :)

**Comment by Jocaru** [comment]
> Okok! i thought `-cenc_decryption_key` was quite old but was added in 5.1 [FFmpeg/FFmpeg@8b64d8d](https://github.com/FFmpeg/FFmpeg/commit/8b64d8d9aa7fede94ab531b7be3f73cd2f3c448c) so maybe you can use some other image like https://github.com/linuxserver/docker-ffmpeg that seem be ubuntu-based and have 5.1+.

Thank, I'll try that! 

I think you can close this issue already, or if you want to keep it open until that new musl version comes out, up to you!

**Comment by wader** [comment]
@Jocaru hey again, master should support dns over tcp now

**Comment by Jocaru** [comment]
> @Jocaru hey again, master should support dns over tcp now

Thank you!
